broadinstitute/single_cell_portal_core
# Speed up plots, remove obsolete code (SCP-3134)

This speeds up server processing time for scatter plots by 30-35%, and removes over 8000 lines of code obsolesced by the React Explore refactor.  

The faster speed comes from removing `text` arrays in plot response data, which had previously been used for tooltips shown upon hovering over scatter plot points.  Performance data below details optimization results.

The removed code was for traditional controllers and ERB templates, which were recently migrated to API controllers and React.  In addition to simply deleting code, these changes include refactoring to eliminate globals and shift around vestigial code (to ease newly-ticketed TODOs).  Our legacy approach with Morpheus is among our largest remaining tech debt on that front -- a brief investigation concluded refactoring that would be non-trivial.

## To test
Verify that the [manual tests for Explore](https://docs.google.com/document/d/1rKcOQPtkbhZCbTqxTnaT-4E29VLzkJrm0kAoWjLx3QA/edit#heading=h.byz5rp8c02bf) pass, given the extensive code updates for the Explore tab in Study Overview.

## Performance data
In tables below for scatter plot endpoints, "Before" and "After" show the response's (compressed) transfer size (e.g. 142 kB), its (uncompressed) resource size (e.g. 923 kB), and its total time between request start and last byte received in response.  "Before" was measured using `development` code.  "Change" shows absolute and [percentage change](https://en.wikipedia.org/wiki/Relative_change_and_difference#Percentage_change).

Tests were local on a real study ("Small intestinal epithelium"; [staging](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP41/small-intestinal-epithelium), [production](https://singlecell.broadinstitute.org/single_cell/study/SCP44/small-intestinal-epithelium)) with 29k cells and 17k genes.  Network speed was [throttled](https://developer.chrome.com/docs/devtools/network/#throttle) to 124 MB/s bandwidth (down, up), 100 ms latency.

Cavaets: this test was on a local SCP instance, and for a relatively small dataset.  

### Reference scatter plot
API endpoint: `clusters?_default`
&nbsp; | Transferred size | Resource size | Time
---|---|---|---
*Before* | 142 kB | 923 kB | 5.92 s
*After* | 96.3 kB | 379 kB | 3.83 s
*Change* | -45.7 kB (32%) | -544 kB (59%) | **-2.09 s (35%)**

### Gene expression scatter plot
API endpoint: `clusters/?_default?gene=Gad2`
&nbsp; | Transferred size | Resource size | Time
---|---|---|---
*Before* | 144 kB | 906 kB | 4.78 s
*After* | 92.3 kB | 406 kB | 3.35 s
*Change* | -51.7 kB (36%) | 500 kB (55%) | **-1.43 s (30%)**

This satisfies SCP-3134.  (Apologies for such a big, dual-purpose PR!)
